.changeset/wicked-squids-do.md

```diff
@@ -0,0 +1,131 @@
+    bytes data
+  );
+
+  interface IStore {
+    function spliceStaticData(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+      uint48 start,
+  -   uint40 deleteCount,
+      bytes calldata data
+    ) external;
+  }
+
+  ```
+
+- All methods that are only valid for dynamic fields (`pushToField`, `popFromField`, `updateInField`, `getFieldSlice`)
+  have been renamed to make this more explicit (`pushToDynamicField`, `popFromDynamicField`, `updateInDynamicField`, `getDynamicFieldSlice`).
+
+  Their `fieldIndex` parameter has been replaced by a `dynamicFieldIndex` parameter, which is the index relative to the first dynamic field (i.e. `dynamicFieldIndex` = `fieldIndex` - `numStaticFields`).
+  The `FieldLayout` parameter has been removed, as it was only used to calculate the `dynamicFieldIndex` in the method.
+
+  ```diff
+  interface IStore {
+  - function pushToField(
+  + function pushToDynamicField(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+  -   uint8 fieldIndex,
+  +   uint8 dynamicFieldIndex,
+      bytes calldata dataToPush,
+  -   FieldLayout fieldLayout
+    ) external;
+
+  - function popFromField(
+  + function popFromDynamicField(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+  -   uint8 fieldIndex,
+  +   uint8 dynamicFieldIndex,
+      uint256 byteLengthToPop,
+  -   FieldLayout fieldLayout
+    ) external;
+
+  - function updateInField(
+  + function updateInDynamicField(
+      ResourceId tableId,
+      bytes32[] calldata keyTuple,
+  -   uint8 fieldIndex,
+  +   uint8 dynamicFieldIndex,
+      uint256 startByteIndex,
+      bytes calldata dataToSet,
+  -   FieldLayout fieldLayout
+    ) external;
+
+  - function getFieldSlice(
+  + function getDynamicFieldSlice(
+      ResourceId tableId,
+      bytes32[] memory keyTuple,
+  -   uint8 fieldIndex,
+  +   uint8 dynamicFieldIndex,
+  -   FieldLayout fieldLayout,
+      uint256 start,
+      uint256 end
+    ) external view returns (bytes memory data);
+  }
+  ```
+
+- `IStore` has a new `getDynamicFieldLength` length method, which returns the byte length of the given dynamic field and doesn't require the `FieldLayout`.
+  
+  ```diff
+  IStore {
+  + function getDynamicFieldLength(
+  +   ResourceId tableId,
+  +   bytes32[] memory keyTuple,
+  +   uint8 dynamicFieldIndex
+  + ) external view returns (uint256);
+  }
+
+- `IStore` now has additional overloads for `getRecord`, `getField`, `getFieldLength` and `setField` that don't require a `FieldLength` to be passed, but instead load it from storage.
+
+- `IStore` now exposes `setStaticField` and `setDynamicField` to save gas by avoiding the dynamic inference of whether the field is static or dynamic.
+
+- The `getDynamicFieldSlice` method no longer accepts reading outside the bounds of the dynamic field.
+  This is to avoid returning invalid data, as the data of a dynamic field is not deleted when the record is deleted, but only its length is set to zero.
```

Comment:
I'm curious about the gas difference between actually deleting and setting length to zero - isn't there a gas return on storage slot deletion?

Reply:
I thought so too and actually implemented the actual removal today. Surprisingly setting it to 0 was _a lot_ more expensive, and the cost grew linearly. Eventually I realized that you still have to provide the entire gas for the transaction to complete, and _at the end_ of the tx receive up to 20% refund.

Reply:
See https://github.com/latticexyz/mud/pull/1581/commits/004fe9ce9ec3012429a498de0a121a32b9639841 and the following 3 commits

Reply:
sorry ethereum for the state bloat we're about to create

Reply:
that's what ethereum gets for being so stingy with the refund